### PR TITLE
Add support for softmax w/ test and benches.

### PR DESCRIPTION
**Pull request by ehiggs**
Softmax for native backends.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request
- Commit: 8fb31e9f04c7ff78302fbf75c94a3bfc0f8b0e1e
  - Commits must be in the following format: **%{type}/%{scope}: %{description}**
  - Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/autumnai/collenchyma-nn/blob/master/CONTRIBUTING.md#git-commit-guidelines

---

This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request
- Commit: 8944cb2b0448e4091c0f72be5ee856ff756118e2
  - Commits must be in the following format: **%{type}/%{scope}: %{description}**

Guidelines are available at https://github.com/autumnai/collenchyma-nn/blob/master/CONTRIBUTING.md#git-commit-guidelines

---

This message was auto-generated by https://gitcop.com

File: src/frameworks/native/helper.rs [hunk 1]
```diff
@@ -371,8 +383,10 @@ macro_rules! impl_ops_softmax_for {
                 x_diff: &mut ::co::tensor::SharedTensor<$t>,
                 result_diff: &mut ::co::tensor::SharedTensor<$t>
             ) -> Result<(), ::co::error::Error> {
-                unimplemented!();
-                Ok(())
+                match x.add_device(self.device()) { _ => try!(x.sync(self.device())) }
+                match x_diff.add_device(self.device()) { _ => try!(x.sync(self.device())) }
```
Comment (MichaelHirn) [inline src/frameworks/native/helper.rs]:
``` rust
match x_diff.add_device(self.device()) { _ => try!(x.sync(self.device())) }
```

syncs, the wrong tensor; should be

``` rust
match x_diff.add_device(self.device()) { _ => try!(x_diff.sync(self.device())) }
```

**Comment by MichaelHirn** [comment]
I found this issue with the wrong sync to be present as well on the activation operations. You could either address them as well in this PR or I fix it up after the merge, before the 0.2.1 release.

I think the general problem is, that the gradients just aren't very nicely implemented. We explore a solution at the moment for this. Interim solution could be to provide a macro, maybe something like this

``` rust
manage_mem!(x, x_diff, result)
```

Anyways, the Softmax looks good to me. Great implementation, thank you very much :+1: 
And the benches are also pretty great, I think

**Comment by ehiggs** [comment]
> I found this issue with the wrong sync to be present as well on the activation operations. You could either address them as well in this PR

Should be fixed now. I didn't put in failing tests to catch them, though.

> I think the general problem is, that the gradients just aren't very nicely implemented. 

I agree that unwrapping the data is pretty tedious. All the branching is especially tedious when we consider the hoops you're jumping through to get everything type safe.

**Comment by MichaelHirn** [comment]
@homu r+

**Pull request merged**